Translate "FAQ: Versioning Policy" (#171)
reactjs/ru.reactjs.org: content/docs/faq-versioning.md

---
id: faq-versioning
title: Политика версионирования
permalink: docs/faq-versioning.html
layout: docs
category: FAQ
---

React следует принципам [семантического версионирования (semver)](https://semver.org/lang/ru/).

Это значит, что для номера версии вида **x.y.z**:

> * Когда мы выпускаем **обратно несовместимые изменения**, мы делаем **мажорный релиз** изменяя число **x** версии  (например, с 15.6.2 до 16.0.0).

Reviewer: ```suggestion
* При выпуске **обратно несовместимых изменений**, мы делаем **мажорный релиз**, изменяя число **x**  (например, с 15.6.2 до 16.0.0).
```

Reviewer: Может немного сократить

> * Когда мы выпускаем **новые возможности**, мы делаем **минорный релиз** изменяя число **y** версии (например, с 15.6.2 до 15.7.0).

Reviewer: ```suggestion
* При выпуске **новых возможностей**, мы делаем **минорный релиз**, изменяя число **y** (например, с 15.6.2 до 15.7.0).
```

> * Когда мы выпускаем **исправления ошибок**, мы делаем **патч-релиз** изменяя число **z** версии (например, с 15.6.2 до 15.6.3).

Reviewer: ```suggestion
* При выпуске **исправлений ошибок**, мы делаем **патч-релиз**, изменяя число **z** (например, с 15.6.2 до 15.6.3).
```

> Мажорные релизы могут содержать новые возможности, также каждый релиз может содержать исправления ошибок.

Reviewer: Кажется, лучше явно разделить.
```suggestion
Мажорные релизы могут содержать новые возможности. Каждый релиз может содержать исправления ошибок.
```

> Мажорные релизы также могут содержать новые возможности, и каждый релиз может содержать исправления ошибок.

Reviewer: ```suggestion
Мажорные релизы могут содержать новые возможности, также каждый релиз может содержать исправления ошибок.
```

### Обратно несовместимые изменения {#breaking-changes}

> Обратно несовместимые изменения неудобны для всех, поэтому мы стараемся минимизировать количество мажорных релизов. Например, React 15 был выпущен в апреле 2016, а React 16 в сентябре 2017. React 17 ожидается не раньше 2019-го года.

Reviewer 2: ```suggestion
Обратно несовместимые изменения неудобны для всех, поэтому мы стараемся минимизировать количество мажорных релизов. Например, React 15 был выпущен в апреле 2016 года, а React 16 — в сентябре 2017 года. React 17 ожидается не раньше 2019 года.
```

> Обратно несовместимые изменения неудобны всем, поэтому мы стараемся минимизировать количество мажорных релизов – например, React 15 был выпущен в апреле 2016-го года, а React 16 был выпущен в сентябре 2017-го года; React 17 ожидается не раньше 2019-го года.

Reviewer: ```suggestion
Обратно несовместимые изменения неудобны для всех, поэтому мы стараемся минимизировать количество мажорных релизов. Например, React 15 был выпущен в апреле 2016, а React 16 в сентябре 2017. React 17 ожидается не раньше 2019-го года.
```

Reviewer: Лучше «для всех», чтобы речь шла про людей, а не изменения.

Вместо этого мы выпускаем новые возможности в минорных релизах. Это значит, что минорные релизы зачастую более интересны, чем мажорные, несмотря на порядковый номер версии.

### Ответственное отношение к стабильности {#commitment-to-stability}

> Когда мы меняем React, мы стараемся минимизировать усилия, необходимые для использования новых возможностей. Когда возможно, мы стараемся сохранить работу старых API, иногда для этого их требуется вынести в отдельный пакет. Например, [примеси были объявлены "вне закона" несколько лет назад](/blog/2016/07/13/mixins-considered-harmful.html) но они до сих пор поддерживаются [через create-react-class](/docs/react-without-es6.html#mixins) и многие проекты продолжают использовать их в стабильном или неподдерживаемом коде.

Reviewer: ```suggestion
Изменяя React, мы стараемся упростить изучение новых возможностей. Кроме этого, мы стараемся сохранить работу старых API, даже если требуется их перенос в отдельный пакет. Например, [мы отказались от примесей несколько лет назад](/blog/2016/07/13/mixins-considered-harmful.html), но они до сих пор поддерживаются [через create-react-class](/docs/react-without-es6.html#mixins) и многие проекты продолжают их использовать в стабильном, устаревшем коде.
```

Reviewer: Попытался упростить: «минимизировать усилия, необходимые для использования» -> «упростить»

Author: Да, отлично получилось. Мне это предложение самому не нравилось )

> React используют больше миллиона разработчиков, поддерживая миллионы компонентов. Только в кодовой базе Facebook более 50 000 React компонентов. Всё это обязывает нас делать обновления до новых версий как можно проще. Если мы не предоставим возможности для обновления, люди застрянут на старых версиях. Мы тестируем наши *пути обновления* прямо в Facebook -- если наша команда из 10 человек может обновить более 50 тысяч компонентов, мы думаем, что с этим справятся и другие React-разработчики. Во многих случаях мы пишем [скрипты автоматизации](https://github.com/reactjs/react-codemod) для обновления синтаксиса компонентов, которые мы выкладываем в открытый доступ для всеобщего использования.

Reviewer 2: ```suggestion
React используют больше миллиона разработчиков, поддерживая миллионы компонентов. Только в кодовой базе Facebook более 50 000 React-компонентов. Всё это обязывает нас делать обновления до новых версий как можно проще. Если мы не предоставим возможности для обновления, люди застрянут на старых версиях. Мы тестируем наши *пути обновления* прямо в Facebook -- если наша команда из 10 человек может обновить более 50 тысяч компонентов, мы думаем, что с этим справятся и другие React-разработчики. Во многих случаях мы пишем [скрипты автоматизации](https://github.com/reactjs/react-codemod) для обновления синтаксиса компонентов, которые мы выкладываем в открытый доступ для всеобщего использования.
```

> React используют больше миллиона разработчиков, поддерживая миллионы компонентов. Только в кодовой базе Facebook более 50 000 React-компонентов. Всё это обязывает нас делать обновления до новых версий как можно проще. Если мы не предоставим возможности для обновления, люди застрянут на старых версиях. Мы тестируем наши *пути обновления* прямо в Facebook -- если наша команда из 10 человек может обновить более 50 тысяч компонентов, мы думаем, что с этим справятся и другие React-разработчики. Во многих случаях мы пишем [скрипты автоматизации](https://github.com/reactjs/react-codemod) для обновления синтаксиса компонентов, которые мы выкладываем в открытый доступ для всеобщего использования.

Reviewer 3: ```suggestion
Больше миллиона разработчиков используют React, поддерживая миллионы компонентов. Только в кодовой базе Facebook более 50 000 React-компонентов. Всё это обязывает нас делать обновления до новых версий как можно проще. Если мы не предоставим возможности для обновления, люди застрянут на старых версиях. Мы тестируем наши *пути обновления* прямо в Facebook -- если наша команда из 10 человек может обновить более 50 тысяч компонентов, мы думаем, что с этим справятся и другие React-разработчики. Во многих случаях мы пишем [скрипты автоматизации](https://github.com/reactjs/react-codemod) для обновления синтаксиса компонентов, которые мы выкладываем в открытый доступ для всеобщего использования.
```

Reviewer 3: > Больше миллиона разработчиков используют React, поддерживая миллионы компонентов.

Может быть так чуть лучше? Хоть немного разнести два слова "миллионы" друг от друга, а то слух режет.
Не настаиваю, впрочем.

Author: Ох, два suggest'а на одну строку, можно принять либо одно, либо другое. Ок, сделаю нормальный commit

Reviewer 2: Отлично и сливаем тогда.

Reviewer 3: @DmitryMarkov 👍

> React используют больше миллиона разработчиков, поддерживая миллионы компонентов. Только в кодовой базе Facebook более 50 000 React-компонентов. Всё это обязывает нас делать обновления до новых версий как можно проще. Если мы не предоставим возможности для обновления, люди застрянут на старых версиях. Мы тестируем наши *пути обновления* прямо в Facebook -- если наша команда из 10 человек может обновить более 50 тысяч компонентов, мы думаем, что с этим справятся и другие React-разработчики. Во многих случаях мы пишем [скрипты автоматизации](https://github.com/reactjs/react-codemod) для обновления синтаксиса компонентов, которые мы выкладываем в открытый доступ для всеобщего использования.

Reviewer 3: ```suggestion
React используют больше миллиона разработчиков, поддерживая миллионы компонентов. Только в кодовой базе Facebook более 50 000 React-компонентов. Всё это обязывает нас делать обновления до новых версий как можно проще. Если мы не предоставим возможности для обновления, люди застрянут на старых версиях. Мы тестируем наши *пути обновления* прямо в Facebook -- если наша команда из 10 человек может обновить более 50 тысяч компонентов, мы думаем, что с этим справятся и другие React-разработчики. Во многих случаях для обновления синтаксиса компонентов мы пишем [скрипты автоматизации](https://github.com/reactjs/react-codemod), которые выкладываем в открытый доступ для всеобщего использования.
```

Reviewer 3: > Во многих случаях для обновления синтаксиса компонентов мы пишем скрипты автоматизации, которые выкладываем в открытый доступ для всеобщего использования.

Поменял порядок слов немного, а то получается, что не скрипты, а компоненты выкладываются в открытый доступ.

> React используют больше миллиона разработчиков, поддерживая миллионы компонентов. Только в кодовой базе Facebook более 50 000 React компонентов. Всё это обязывает нас делать обновления до новых версий как можно проще. Если мы не предоставим возможности для обновления, люди застрянут на старых версиях. Мы тестируем наши *пути обновления* прямо в Facebook -- если наша команда из 10 человек может обновить более 50 тысяч компонентов, мы думаем, что с этим справятся и другие React-разработчики. Во многих случаях мы пишем [скрипты автоматизации](https://github.com/reactjs/react-codemod) для обновления синтаксиса компонентов, которые мы выкладываем в открытый доступ для всеобщего использования.

Reviewer: ```suggestion
Больше миллиона разработчиков использую React, поддерживая миллионы компонентов. Только в кодовой базе Facebook более 50 000 React-компонентов. Всё это обязывает нас делать обновления до новых версий как можно проще. Если мы сделаем большие изменения, не предоставив возможности для обновления, люди застрянут на старых версиях. Мы тестируем наши *пути обновления* прямо в Facebook -- если наша команда из 10 человек может обновить более 50 тысяч компонентов, мы думаем, что с этим справятся и другие React-разработчики. Во многих случаях мы пишем [codemod-скрипты](https://github.com/reactjs/react-codemod) для обновления синтаксиса компонентов, которые мы выкладываем в открытый доступ для всеобщего использования.
```

### Постепенное обновление через предупреждения {#gradual-upgrades-via-warnings}

> Режим для разработчиков в React включает множество полезных предупреждений. Когда возможно, мы добавляем предупреждения для будущих обратно несовместимых изменений. Таким образом, если ваше приложение не показывает предупреждений в консоли в последнем релизе, значит оно готово к следующей мажорной версии. Это позволяет вам обновлять ваше приложение компонент за компонентом по одиночке.

Reviewer 2: ```suggestion
Версии разработки React включает множество полезных предупреждений. Когда возможно, мы добавляем предупреждения для будущих обратно несовместимых изменений. Таким образом, если ваше приложение не показывает предупреждений в консоли в последнем релизе, значит оно готово к следующей мажорной версии. Это позволяет вам обновлять приложение компонент за компонентом по одиночке.
```

Author: Может, не версия, а сборка? В любом случае надо согласовать "Версия включает" или "Версии включают"

Reviewer 2: Согласен, изменил.

Reviewer: Может лучше _Сборки в режиме разработки_ или _Сборки проекта в режиме разработки_, а то кажется, что собираем разработку, а не проект. 

_включают_, если оставим множественное число.

> Режим для разработчиков в React включает множество полезных предупреждений. Когда возможно, мы добавляем предупреждения для будущих обратно несовместимых изменений. Таким образом, если ваше приложение не показывает предупреждений в консоли в последнем релизе, значит оно готово к следующей мажорной версии. Это позволяет вам обновлять ваше приложение компонент за компонентом по одиночке.

Reviewer: ```suggestion
Режим для разработчиков в React включает множество полезных предупреждений. По возможности, мы добавляем предупреждения для будущих обратно несовместимых изменений. Таким образом, если ваше приложение не показывает предупреждений в консоли в последнем релизе, значит оно готово к следующей мажорной версии. Это позволяет вам обновлять ваше приложение компонент за компонентом по одиночке.
```

> Предупреждения в режиме разработчика не показываются во времени выполнения вашего приложения. Таким образом, вы можете быть уверены, что ваше приложение будет вести себя одинаково в режиме разработчика и продакшен-режиме. Разница лишь в том, что продакшен сборка не будет показывать предупреждения в консоли, что, к тому же, более производительно. (Если вы вдруг заметили предупреждение в продакшен-режиме, откройте ишью.)

Reviewer 2: ```suggestion
Предупреждения в режиме разработчика не показываются во времени выполнения вашего приложения. Таким образом, вы можете быть уверены, что ваше приложение будет вести себя одинаково в режиме разработчика и продакшен-режиме. Разница лишь в том, что продакшен-сборка не будет показывать предупреждения в консоли, что, к тому же, более производительно. (Если вы вдруг заметили предупреждение в продакшен-режиме, откройте ишью.)
```

> Предупреждения в режиме разработчика не показываются во времени выполнения вашего приложения. Таким образом, вы можете быть уверены, что ваше приложение будет вести себя одинаково в режиме разработчика и продакшен-режиме. Разница лишь в том, что продакшен-сборка не будет показывать предупреждения в консоли, что, к тому же, более производительно. (Если вы вдруг заметили предупреждение в продакшен-режиме, откройте ишью.)

Reviewer 3: ```suggestion
Предупреждения в режиме разработки не влияют на то, как исполняется ваше приложение. Таким образом, вы можете быть уверены, что ваше приложение будет вести себя одинаково в режиме разработки и продакшен-режиме. Разница лишь в том, что продакшен-сборка не будет показывать предупреждения в консоли, что, к тому же, более производительно. (Если вы вдруг заметили предупреждение в продакшен-режиме, откройте ишью.)
```

### Что считается обратно несовместимым изменением? {#what-counts-as-a-breaking-change}

Как правило, мы *не* повышаем мажорную версию для следующих изменений:

> * **Предупреждения для разработчиков.** Поскольку они не затрагивают поведение в продакшен-режиме, мы можем добавлять или изменять существующие предупреждения между мажорными версиями. На самом деле, это позволяет нам доходчиво предупреждать о новых мажорных изменениях.

Reviewer: ```suggestion
* **Предупреждения для разработчиков.** Поскольку они не затрагивают поведение в продакшен-режиме, мы можем добавлять или изменять существующие предупреждения между мажорными версиями. Это позволяет нам заранее предупреждать о новых мажорных изменениях.
```

> * **API с приставкой `unstable_`.** Они поставляются как экспериментальные возможности, в API которых мы не до конца уверены. Выпуская такие возможности с приставкой `unstable_` мы можем обновлять их быстрее и прийти к стабильному API быстрее.

Reviewer: ```suggestion
* **API с приставкой `unstable_`.** Они добавляют экспериментальные возможности, в API которых мы не уверены до конца. Выпуская такие возможности с приставкой `unstable_`, мы можем их обновлять и переходить к стабильному API быстрее.
```

> * **Alpha и canary версии React.** Мы предоставляем alpha версии React как способ как можно скорей попробовать новые возможности, но нам нужна гибкость для внесения изменений на основе обратной связи, полученной в период альфа-тестирования. Если вы используете такие версии, то имейте ввиду, что API могут поменяться в стабильной версии.

Reviewer 2: Нужно поработать ещё над этим (убрать двойное "как")

```suggestion
* **Альфа и канареечные версии React.** Мы предоставляем альфа-версии React как способ как можно скорей попробовать новые возможности, но нам нужна гибкость для внесения изменений на основе обратной связи, полученной в период альфа-тестирования. Если вы используете такие версии, то имейте в виду, что API могут поменяться в стабильной версии.
```

> * **Альфа и канареечные версии React.** Мы предоставляем альфа-версии React как способ как можно скорей попробовать новые возможности, но нам нужна гибкость для внесения изменений на основе обратной связи, полученной в период альфа-тестирования. Если вы используете такие версии, то имейте в виду, что API могут поменяться в стабильной версии.

Reviewer: ```suggestion
* **Альфа и канареечные версии React.** Альфа-версии React позволяют попробовать новые возможности раньше. Мы можем вносить в них изменения на основе обратной связи, полученной в период альфа-тестирования. Если вы используете такие версии, имейте в виду, что API может измениться в стабильной версии.
```

> * **Недокументированные API и внутренние структуры данных.** Если вы полагаетесь на внутренние переменные типа `__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED` или `__reactInternalInstance$uk43rzhitjg`, вы не получаете никаких гарантий. Это ваша ответственность.

Reviewer: ```suggestion
* **Недокументированные API и внутренние структуры данных.** Использование внутренних переменных, таких как `__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED` или `__reactInternalInstance$uk43rzhitjg` не даёт никаких гарантий.
```

Author: @gcor может так: `Мы снимаем ответственность за использование вами внутренних переменных, таких как `__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED` или `__reactInternalInstance$uk43rzhitjg` .`

Reviewer: Тогда надо уточнять с кого ответственность снимается: _Мы снимаем с себя ответственность_. Но это звучит слишком странно, мне кажется, как-то с претензией. Ну и странно, если кто-то кроме меня несёт ответственность за мой же код.

Мне всё равно кажется, что отсутствие гарантий звучит мягче.

Reviewer 2: _Мы не гарантируем работоспособность кода в случае использования...._?

Reviewer 3: > Это ваша ответственность.

Немножечко коряво -- не ясно, что "_это_".
В русском есть _действуете на свой страх и риск_, например. Или можно сказать как-то проще, типа _будьте готовы к проблемам_.

> Эта политика создана, чтобы помочь на практике. Мы не хотим, чтобы у вас болела голова. Если бы мы поднимали мажорную версию слишком часто, то доставили бы множество проблем всему сообществу. И это бы не позволило улучшать React так быстро, как нам хотелось.

Reviewer 3: ```suggestion
Наша политика разработана, чтобы быть практичной. Мы не хотим создавать вам головную боль. Если бы мы поднимали мажорную версию слишком часто, то доставили бы множество проблем всему сообществу. И это бы не позволило улучшать React так быстро, как нам хотелось.
```

> Эта политика создана быть практичной: безусловно, мы не хотим, чтобы у вас болела голова. Если бы мы поднимали мажорную версию каждый раз во всех вышеперечисленных случаях, мы бы выпускали их столь часто, что доставили бы кучу боли всему сообществу. И вряд ли бы мы не смогли улучшать React столь быстро, как нам хотелось бы.

Reviewer: ```suggestion
Эта политика создана, чтобы помочь на практике. Мы не хотим, чтобы у вас болела голова. Если бы мы поднимали мажорную версию слишком часто, то доставили бы множество проблем всему сообществу. И это бы не позволило улучшать React так быстро, как нам хотелось.
```

> В любом случае, если мы думаем, что изменения могут иметь далекоидущие последствия в сообществе, мы постараемся сделать всё возможное, чтобы предоставить плавный пошаговый переход от старой версии к новой.

Reviewer: ```suggestion
Если мы думаем, что изменения смогут вызвать широкие последствия в сообществе, мы постараемся сделать всё возможное, чтобы предоставить плавный переход от старой версии к новой.
```

Тем не менее, если мы ожидаем, что изменение в этом списке вызовет широкие проблемы в сообществе, мы все равно сделаем все возможное, чтобы обеспечить постепенный путь миграции.

Author: @gcor что такое `широкие последствия`? Может, `глобальные`, если не нравится исходный вариант?

Reviewer: Может попробовать упростить: _вызовут проблемы в сообществе_

Author: _могут вызвать_, кмк

Reviewer: Да, можно так
